docs/datatypes/tmap.md

```diff
@@ -0,0 +1,279 @@
+---
+id: datatypes_tmap
+title: "TMap"
+---
+
+A `TMap[A]` is a mutable map that can participate in transactions in STM.
+
+## Create a TMap
+
+Creating an empty `TMap`:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val emptyTMap: STM[Nothing, TMap[String, Int]] = TMap.empty[String, Int]
+```
+
+Or creating a `TMap` with specified values:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val specifiedValuesTMap: STM[Nothing, TMap[String, Int]] = TMap.make(("a", 1), ("b", 2), ("c", 3))
+```
+
+Alternatively, you can create a `TMap` by providing a collection of tuple values:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val iterableTMap: STM[Nothing, TMap[String, Int]] = TMap.fromIterable(List(("a", 1), ("b", 2), ("c", 3)))
+```
+
+## Put a key-value pair to a TMap
+
+New key-value pair can be added to the map in the following way:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val putElem: UIO[TMap[String, Int]] = (for {
+  tMap <- TMap.make(("a", 1), ("b", 2))
+  _    <- tMap.put("c", 3)
+} yield tMap).commit
+```
+
+Another way of adding an entry in the map is by using `merge`:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val mergeElem: UIO[TMap[String, Int]] = (for {
+  tMap <- TMap.make(("a", 1), ("b", 2), ("c", 3))
+  _    <- tMap.merge("c", 4)((x, y) => x * y)
+} yield tMap).commit
+```
+
+If the key is not present in the map it behaves as a simple `put` method. It merges the existing value with the new one using the provided function otherwise.
+
+## Remove an element from a TMap
+
+The simplest way to remove a key-value pair from `TMap` is using `delete` method that takes key:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val deleteElem: UIO[TMap[String, Int]] = (for {
+  tMap <- TMap.make(("a", 1), ("b", 2), ("c", 3))
+  _    <- tMap.delete("b")
+} yield tMap).commit
+```
+
+Also, it is possible to remove every key-value pairs that satisfies provided predicate:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val removedEvenValues: UIO[TMap[String, Int]] = (for {
+  tMap <- TMap.make(("a", 1), ("b", 2), ("c", 3), ("d", 4))
+  _    <- tMap.removeIf((_, v) => v % 2 == 0)
+} yield tMap).commit
+```
+
+Or you can keep all key-value pairs that match predicate function:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val retainedEvenValues: UIO[TMap[String, Int]] = (for {
+  tMap <- TMap.make(("a", 1), ("b", 2), ("c", 3), ("d", 4))
+  _    <- tMap.retainIf((_, v) => v % 2 == 0)
```

Comment:
Point out the parity between `removeIf` / `retainIf` and `filter` / `filterNot`, as well as with `transform`s and `map`s. This remark applies for sets as well.